shared_stack.bash

```diff
@@ -0,0 +1,214 @@
+##############
+
+set -e
+mkdir -p "$STATE"
+cd "$STATE"
+# Label temporaries with our unique PID
+tmp=tmp.$$
+mkdir -p "$tmp" "$ROOT/tag"
+[ -e tags.deleted ] || touch tags.deleted
+umask 022
+
+# Get list of tags
+$dryrun curl -L https://ls.st/lsstinstall -o $tmp/lsstinstall
+$dryrun mv $tmp/lsstinstall lsstinstall
+curl -L https://eups.lsst.codes/stack/src/tags | while read -r line; do
+  [[ "$line" =~ \>([dvw][0-9_]+(rc[0-9]+)?)\.list\< ]] && echo "${BASH_REMATCH[1]}"
+done | sort > $tmp/tags
+
+for tag in $(comm -1 -3 <(sort tags.deleted) $tmp/tags); do
+  dir="$ROOT/tag/$tag"
+  [ -d "$dir" ] && continue
+  $dryrun mkdir -p "${dir}.$tmp" || continue
+  (
+     set -e
+     $dryrun cd "${dir}.$tmp"
+     $dryrun bash "$STATE/lsstinstall" -T "$tag" -p "$ROOT"/conda -S
+     set +x
+     $dryrun source loadLSST.sh
+     set -x
+     # Install the developer add-on packages if needed
+     # When rubin-env-developer becomes available, replace with this:
+     # conda list --json | grep rubin-env-developer > /dev/null \
+     #   || mamba install rubin-env-developer
+     if [ -n "$dryrun" ]; then
+       echo "conda list --json | grep mypy > /dev/null || mamba install ..."
+     else
+       # shellcheck disable=SC2046
+       conda list --json | grep mypy > /dev/null \
+         || mamba install -y -c conda-forge --no-update-deps $(cat "$STATE/developer.txt")
+     fi
+     $dryrun eups distrib install -t "$tag" "$PRODUCT"
+
+     # Rewrite load scripts and write useful files
+     $dryrun cp loadLSST.sh tmp.loadLSST.sh
+     if [ -n "$dryrun" ]; then
+       echo "echo . $ROOT/loadLSST.sh ... > loadLSST.sh"
+       echo "echo conda activate ... > switch.sh"
+       echo "echo ... > env_name"
+     else
+       echo "LSST_CONDA_ENV_NAME=$LSST_CONDA_ENV_NAME . \"$ROOT/loadLSST.sh\"" > loadLSST.sh
+       echo "conda activate $LSST_CONDA_ENV_NAME" > switch.sh
+       echo "$LSST_CONDA_ENV_NAME" > env_name
+     fi
+  ) && [ ! -d "$dir" ] && $dryrun mv "${dir}.$tmp" "$dir"
+  # Clean up if we failed or lost the race
+  [ -d "${dir}.$tmp" ] && rm -rf "${dir}.$tmp"
+done
+rm $tmp/tags
+
+# List the installed tags
+(
+  cd "$ROOT/tag"
+  ls -dr d_* > "$STATE/$tmp/d_list" || true
+  ls -dr w_* > "$STATE/$tmp/w_list" || true
+)
+
+# Make the latest links if needed
+for type in d w; do
+  [ -f "$tmp/${type}_list" ] || continue
+  latest=$(head -n 1 "$tmp/${type}_list")
+  link="$ROOT/${type}_latest"
+  if [[ ! -L "$link" || "$(readlink "$link")" != "$ROOT/tag/$latest" ]]; then
+    $dryrun rm -f "$link" && $dryrun ln -s "$ROOT/tag/$latest" "$link"
+  fi
+done
+# Update loadLSST.sh if needed
+tmp_load="$ROOT/w_latest/tmp.loadLSST.sh"
+if [ -n "$dryrun" ]; then
+  echo "[ -f \"$tmp_load\" ] && mv \"$tmp_load\" \"$ROOT/loadLSST.sh\""
+elif [ -f "$tmp_load" ] && ! diff -q "$ROOT/loadLSST.sh" "$tmp_load" > /dev/null 2>&1; then
+  [ -f "$ROOT/loadLSST.sh" ] && mv "$ROOT/loadLSST.sh" "$ROOT/loadLSST.sh.bak"
+  mv "$tmp_load" "$ROOT/loadLSST.sh"
```

Comment:
Sorry, this bit here is what has been holding me up, I confess it was not strait forward to me how tied into the above.

Reply:
This is always taking the loadLSST.sh from the latest weekly.  As mentioned above, this should change very rarely, but I wanted it to automatically update if needed.

Note that loadLSST.sh is created by lsstinstall, so it is not versioned along with the rest of the stack.

Reply:
That is what I was worried about, if somehow it did get updated, and somehow broke a lot of old stack versions without us realizing (as you say its unlikely, but I wanted to highlight)